nanshe_ipython.ipynb

```diff
@@ -471,6 +474,257 @@
       "cell_type": "markdown",
       "metadata": {},
       "source": [
+        "### Registration"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "def find_best_match(matches):\n",
+        "    i = numpy.argmin((matches ** 2).sum(axis=0))\n",
+        "\n",
+        "    return matches[:, i]\n",
+        "\n",
+        "\n",
+        "def compute_offset(match_mask):\n",
+        "    frame_shape = np.array(match_mask.shape)\n",
+        "    half_frame_shape = frame_shape // 2\n",
+        "\n",
+        "    matches = np.array(match_mask.nonzero())\n",
+        "    if matches.size == 0:\n",
+        "        matches = np.array([[0], [0]])\n",
+        "\n",
+        "    above = (matches > half_frame_shape[:, None]).astype(matches.dtype)\n",
+        "    matches -= above * frame_shape[:, None]\n",
+        "\n",
+        "    return find_best_match(matches)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "%%time\n",
+        "\n",
+        "\n",
+        "num_reps = 5\n",
+        "tmpl_hist_wght = 0.25\n",
+        "thld_rel_dist = 0.0\n",
+        "\n",
+        "block_frames = 1\n",
+        "norm_frames = 100\n",
+        "\n",
+        "\n",
+        "# Somehow we can't overwrite the file in the container so this is needed.\n",
+        "io_remove(data_basename + postfix_reg + zarr_ext)\n",
+        "io_remove(data_basename + postfix_reg + h5_ext)\n",
+        "\n",
+        "\n",
+        "with open_zarr(data_basename + postfix_dn + zarr_ext, \"r\") as f:\n",
+        "    with get_executor(client) as executor:\n",
+        "        # Load and prep data for computation.\n",
+        "        imgs = f[\"images\"]\n",
+        "        da_imgs = da.from_array(imgs, chunks=imgs.chunks)\n",
+        "        da_imgs = da_imgs.rechunk((block_frames,) + imgs.shape[1:])\n",
+        "\n",
+        "        da_imgs = da_imgs[1:-1]\n",
```

Comment:
I don't think this should be here as it is denoising already. That said, it probably should live in its own step so it is clear what is going on.